socks5.c

```diff
@@ -0,0 +1,337 @@
+        conn_len = 22;
+   } else {
+        // assume a resolvable domain name
+        // previously validate to be 0 < x <= 255 bytes
+        uint8_t alen = (uint8_t)strlen(addr);
+        conn[3] = 0x03; // ATYP = DOMAINNAME
+        conn[4] = alen;
+        memcpy((void *)(conn+ 5), (void *)addr, alen);
+        conn[5+alen] = port >> 8;
+        conn[6+alen] = port & 0xFF;
+        conn_len = 7 + alen;
+    }
+    int err = dill_bsend(obj->u, conn, conn_len, deadline);
+    if (dill_slow(err)) {return -1;}
+    // reply has the same form/length limit, reuse request as reply buffer
+    // read first 5 octets (through ATYP) to determine how much more to read
+    err = dill_brecv(obj->u, conn, 5, deadline);
+    if (dill_slow(err)) {return -1;}
+    switch (conn[3]) {
+        case 0x01:
+            // IPv4, already read 1 bytes, read remaining 3 and port
+            err = dill_brecv(obj->u, conn + 5, 5, deadline);
+            if (dill_slow(err)) {return -1;}
+            break;
+        case 0x03:
+            // IPv4, already read 1 bytes, read remaining 15 and port
+            err = dill_brecv(obj->u, conn + 5, 17, deadline);
+            if (dill_slow(err)) {return -1;}
+            break;
+        case 0x04:
+            // IPv4, already read 1 bytes, read ALEN-1 add + 2 for port
+            err = dill_brecv(obj->u, conn + 5, conn[4] + 1, deadline);
+            if (dill_slow(err)) {return -1;}
+            break;
+        default:
+            errno = EPROTO;
+            return -1;
+    }
+    if (conn[1] == 0x00) {
+        // success - from here on should treat as if attached via tcp_connect
+        obj->connect = 1;
+        return 0;
+    }
+    // failed, it's just a matter of deciding which error code to return
+    switch (conn[1]) {
+        case 0x01:
+            // general SOCKS server failure
+            errno = EIO;
+            break;
+        case 0x02:
+            // connection not allowed by ruleset
+            errno = EACCES;
+            break;
+        case 0x03:
+            // Network unreachable
+            errno = ENETUNREACH;
+            break;
+        case 0x04:
+            // Host unreachable
+            errno = EHOSTUNREACH;
+            break;
+        case 0x05:
+            // Connection Refused
+            errno = ECONNREFUSED;
+            break;
+        case 0x06:
+            // TTL expired
+            errno = ETIMEDOUT;
+            break;
+        case 0x07:
+            // Command not supported
+            errno = EOPNOTSUPP;
+            break;
+        case 0x08:
+            // Address type not supported
+            errno = EAFNOSUPPORT;
+            break;
+        default:
+            // Unknown error code
+            errno = EPROTO;
+    }
+    return -1;
+}
+
+int dill_socks5_close(int s, int64_t deadline) {
+    struct dill_socks5_sock *obj = dill_hquery(s, dill_socks5_type);
+    if(dill_slow(!obj)) return -1;
+    return dill_tcp_close(obj->u, deadline);
```

Comment:
The underlying socket in not necessarily a TCP socket. It may be TLS or whatever.

Reply:
Good point. In this case I think it's better to force socks5_detach and enable the program to call tcp_close or tls_detach/tcp_close or whatever close is appropriate. Regardless, once connected the proxy becomes completely passive. No addition control is required nor allowed.

(so I've implemented and exposed socks5_detach)